dist/dist_client.go

```diff
@@ -0,0 +1,500 @@
+	req, err := http.NewRequest(http.MethodHead, blobURI.String(), nil)
+
+	if err != nil {
+		return false
+	}
+
+	if o.authCreds != "" {
+		req.Header.Add("Authorization", "Basic "+o.authCreds)
+	}
+
+	resp, err := o.client.Do(req)
+
+	if err != nil {
+		return false
+	}
+
+	defer resp.Body.Close()
+
+	return resp.StatusCode == http.StatusOK
+}
+
+func (o *OciRepo) GetLayer(ldigest string) (io.ReadCloser, int64, error) {
+	name := o.ref.name
+	blobURI, err := url.Parse(fmt.Sprintf("/v2/%s/blobs/%s", name, ldigest))
+
+	if err != nil {
+		return nil, -1, errors.Wrapf(err, "couldn't parse URL for repo %s blob %s", name, ldigest)
+	}
+
+	blobURI = o.url.ResolveReference(blobURI)
+	req, err := http.NewRequest("GET", blobURI.String(), nil)
+
+	if err != nil {
+		return nil, -1, errors.Wrapf(err, "Couldn't create GET request for %v", blobURI)
+	}
+
+	if o.authCreds != "" {
+		req.Header.Add("Authorization", "Basic "+o.authCreds)
+	}
+
+	resp, err := o.client.Do(req)
+
+	if err != nil {
+		return nil, -1, errors.Wrapf(err, "Error getting layer %s", ldigest)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		resp.Body.Close()
+		return nil, -1, fmt.Errorf("bad return code %d getting layer", resp.StatusCode)
+	}
+
+	return resp.Body, -1, err
+}
+
+// StartLayer starts a blob upload session
+func (o *OciRepo) StartLayer() (*url.URL, error) {
+	name := o.ref.name
+	uploadURI, err := url.Parse(fmt.Sprintf("/v2/%s/blobs/uploads/", name))
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to parse upload URL")
+	}
+
+	uploadURI = o.url.ResolveReference(uploadURI)
+	req, err := http.NewRequest("POST", uploadURI.String(), nil)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed opening POST request")
+	}
+
+	if o.authCreds != "" {
+		req.Header.Add("Authorization", "Basic "+o.authCreds)
+	}
+
+	resp, err := o.client.Do(req)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed posting request %v", req)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusAccepted {
+		return nil, fmt.Errorf("server returned an error %d", resp.StatusCode)
+	}
+
+	_, err = ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error reading response body for %s", name)
+	}
+
+	session, err := resp.Location()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed decoding response")
+	}
+
+	return o.url.ResolveReference(session), nil
+}
+
+// @path is the uuid upload path returned by the server to our Post request.
+// @stream is the data source for the layer.
+// Return the digest and size of the layer that was uploaded.
+//nolint (gocognit)
+func (o *OciRepo) CompleteLayer(session *url.URL, stream io.Reader) (digest.Digest, int64, error) {
+	digester := sha256.New()
+	hashReader := io.TeeReader(stream, digester)
+	// using "chunked" upload
+	count := int64(0)
+	for {
+		const maxSize = 10 * 1024 * 1024
+		var buf bytes.Buffer
+		size, err := io.CopyN(&buf, hashReader, maxSize)
```

Comment:
What’s the benefit of using chunked upload instead of a single long-lived request? AFAICS this only adds overhead with no benefit I can see (e.g. I could imagine that chunked upload might allow restarting on partial failure, but this code doesn’t do that).